### ISSUE TITLE
grouping of samples for seq2print

### ISSUE THREAD

**Issue by SHhAaNnE** [buenrostrolab/scPrinter#75]
Thank you for developing and sharing such a powerful tool with scPrinter and seq2print

I wanted to reach out with a question about training the seq2print model and how best to group samples. In the bulk ATAC-seq tutorial, I noticed that a separate model is trained for each sample, which makes sense since each sample there represents a distinct cell type.

In my case, I'm working with nine samples derived from the same cell type, but split across different experimental conditions. For the sake of reducing computation time, I've been considering whether to train a single model across all samples combined, but I was unsure if this would have drastic implications on the quality/interpretability of results and final transcription factor binding scores.

I'd love to hear your thoughts on how you typically decide whether to group or separate samples for model training in seq2print, especially when cell type is constant but conditions vary. 

Thanks in advance for your time and insights.

**Comment by HYsxe** [comment]
Thanks for the question! Below are some suggestions:

(1) When the number of samples is not super high (say single digits) I would usually try to start training a separate model for each sample. This way you get a sense of variation across samples. For example if I have 4 control, 4 treated, then training 8 separate models allows me to see how consistent the difference between control and treatment is. When you collapse replicates and train one model, you no longer get this information and if one of your samples is an outlier it could totally bias your result.

(2) If you have a really large number of samples, say hundreds, then maybe consider doing LoRA to scale up. In this case you need a good way to encode the sample embedding so that the model can capture the differences.

(3) If training takes a long time you can consider starting with only 1 fold and maybe merging similar samples just for testing purposes, but for final publication you prob want to train separate models for each sample.

Let me know if this helps!

**Comment by HYsxe** [comment]
In that case I would recommend starting with what you described and see how it works! Make sure to check if the samples have any significant imbalances in sequencing depth, in which case you need to quantile normalize your TF binding scores per sample at the end. I am curious how this group + sample fintune will work so please let us know how it goes!

**Comment by jtls0n** [comment]
Since these are all bulk ATAC, I'm struggling a bit to manage the grouping for the fine tuning and embedding. 

For the embedding, I initially considered the PCA for chromvar scores (as is used in the PBMC scATAC tutorial), but the groups are quite distinct based on this and almost completely separates the samples by group.

I read this discussion about trying to minimize bias from the embedding in the fine tuning (https://github.com/buenrostrolab/scPrinter/issues/55). One of the goals in my analysis is to explore variability between the samples one of the group, where the ChromVAR scores vary a bit by disease state.

Based on that, I tried to set the grouping by each bulk ATAC sample (where the combined fragments file has sample as barcode):
```
barcodegroups = pd.DataFrame({'barcode': samples, 'group': samples})
cell_grouping, group_names = scp.utils.df2cell_grouping(printer, barcodegroups)
```

then I tried to set the embedding this way (using a dataset with 5 groups and 70 total samples): 
```
categories = np.array(list(range(0, 70))).reshape(-1, 1)
encoder = OneHotEncoder(sparse_output=False)
embeddings = encoder.fit_transform(categories)
```

**Comment by HYsxe** [comment]
I think this make sense. In your case if you use PCA embedding for LoRA, the samples with similar PC embeddings might just get very similar model predictions, so your current design might be more helpful. 

When you do the one-hot encoding, you could also consider adding group-specific onehot dimensions, although we will only know how much it helps after we try.

**Comment by HYsxe** [comment]
Closing this issue now but feel free to reopen if you have more questions! Please don't hesitate to reach out
